### DISCUSSION TITLE
Request: Add cgroup memory checking support

### DISCUSSION THREAD

**Discussion by paultiq** [duckdb/duckdb#7788] [Ideas]
As discussed in https://github.com/duckdb/duckdb/issues/7651:

Currently, DuckDB uses the following code to determine the maximum memory:
> https://github.com/duckdb/duckdb/blob/0d946c04df35e0d2ddfa033ce586dc18671a3c80/src/common/file_system.cpp#L82, 

This approach does not take into account the memory limits imposed by [cgroups](https://man7.org/linux/man-pages/man7/cgroups.7.html) to avoid exceeding specified memory limits and avoid OOMs.

DuckDB already checks cgroups for CPU limits, so it seems consistent to me to check for both memory and cpu: 
https://github.com/duckdb/duckdb/blob/0d946c04df35e0d2ddfa033ce586dc18671a3c80/src/main/config.cpp#L229-L232

**Comment by Tishj** [comment]
For the CGroup CPU limits, testing was the hardest part, am I correct in thinking that `docker run --memory=1g <image_name>` is the only thing needed to set this limit on a docker container? (assuming that the image in question uses a Linux base image)

**Comment by paultiq** [comment]
There's a few different ways to set cgroup, including on individual processes/etc.

Thinking about it, I'm actually not convinced the cpu check is correct, using /sys/fs seems incorrect when checking current limits (I would've expected /proc/self, but this isn't my area of expertise). I think it'd work with a single global cgroup, but not in 'all' cases.

For configuring, I'll see if I can whip up a post-boot test case that creates the cgroups rather than on-host version.

**Comment by paultiq** [comment]
- cgroup v1: 
`docker run -it -m=200m ubuntu cat /sys/fs/cgroup/memory/memory.limit_in_bytes`
> 209715200

- cgroup v2 (after setting `cgroup_no_v1=all`)
`docker run -it -m=250m ubuntu cat /sys/fs/cgroup/memory.max`
> 262144000

Root cgroup Only:
 I think it's worth noting (documenting) that both of these checks, and the CPU check, only check the system/root cgroup, and 
doesn't account for additional process specific cgroups. To get the current processes cgroup, you have to look at /proc/self/cgroup... but this isn't something I used in practice, and don't know how common it is.

**Comment by paultiq** [comment]
How other modules check CPU counts. 

- [DuckDB](https://github.com/duckdb/duckdb/blob/8c32403411d628a400cc32e5fe73df87eb5aad7d/src/main/config.cpp#LL281C6-L281C6): Checks `std::thread::hardware_concurrency()` then root cgroup
- [PyArrow](https://github.com/apache/arrow/blob/f44f7685638e4afb97f8d6d536d8538b5292b47d/cpp/src/arrow/util/thread_pool.cc#LL517C1-L517C1): Checks OMP_NUM_THREADS, then `std::thread::hardware_concurrency()`. Does not check cgroup
- [Polars uses num_cpus crate](https://github.com/seanmonstar/num_cpus/blob/master/src/linux.rs): This seems like the most robust version. Notably, it checks /proc/self/cgroup, etc.

The num_cpus crate seemed pretty interesting. I didn't (yet) find a good example for memory checking, Polars, for instance, didn't seem to do much more than check system memory:  [memory.rs](https://github.com/pola-rs/polars/blob/main/polars/polars-lazy/polars-pipe/src/executors/sinks/memory.rs)

Edit: 
- Java seems to have the most robust logic here for both CPU and memory [CGroup detection via -XX:+UseContainerSupport](https://github.com/openjdk/jdk/blob/ecb17532dc8f3e271ad2d6550127a2253569cf9b/src/java.base/linux/classes/jdk/internal/platform/CgroupInfo.java#L29) that I can find. for more info: https://bugs.openjdk.org/browse/JDK-8230305

**Comment by soerenwolfers** [comment]
https://github.com/duckdb/duckdb/pull/16608 should fix a lot of issues in this area

**Reply by Tishj** [reply]
The last pieces of the puzzle here is using and checking `mountinfo`, I might revisit this at some point.
For now we fall back to the root directory if all else fails - as a way to simulate the most common usecase for `mountinfo`

**Comment by paultiq** [comment]
Agree, just checked my k8s's... memory_limit and worker_threads is correct in my environment (using the nightly). 🎉 

Thx!
